File: backend/src/main/java/com/saif/portfolio/service/AuthService.java
package com.saif.portfolio.service;

import java.util.Map;

import com.saif.portfolio.dto.ForgotPasswordRequest;
import com.saif.portfolio.dto.LoginRequest;
import com.saif.portfolio.dto.RegisterRequest;
import com.saif.portfolio.dto.ResetPasswordRequest;
import com.saif.portfolio.dto.TokenResponse;
import com.saif.portfolio.dto.VerifyEmailRequest;

public interface AuthService {
    void register(RegisterRequest request);
    void verifyEmail(VerifyEmailRequest request);
    void resendEmailVerificationOtp(String email);
    TokenResponse login(LoginRequest request, String ip);
    TokenResponse refreshToken(String refreshToken);
    void forgotPassword(ForgotPasswordRequest request);
    void resetPassword(ResetPasswordRequest request);
    void logout(String refreshToken);
    void logoutAllOtherDevices(String currentRefreshToken);
    TokenResponse socialLogin(String provider, String providerUserId, Map<String, Object> profile);
}

File: frontend/src/services/authService.ts
import { apiClient } from "@/lib/apiClient";
import type {
  ApiResponse,
  LoginRequest,
  RegisterRequest,
  VerifyEmailRequest,
  ForgotPasswordRequest,
  ResetPasswordRequest,
  ResendOtpRequest,
  TokenResponse,
  User,
} from "@/types/auth";

export const authService = {
  async register(data: RegisterRequest): Promise<ApiResponse<void>> {
    const res = await apiClient.post<ApiResponse<void>>("/auth/register", data);
    return res.data;
  },

  async login(data: LoginRequest): Promise<TokenResponse> {
    const res = await apiClient.post<ApiResponse<TokenResponse>>(
      "/auth/login",
      data
    );
    const tokens = res.data.data;
    localStorage.setItem("accessToken", tokens.accessToken);
    return tokens;
  },

  async verifyEmail(data: VerifyEmailRequest) {
    return (await apiClient.post<ApiResponse<void>>("/auth/verify-email", data))
      .data;
  },

  async resendVerifyOtp(data: ResendOtpRequest) {
    return (
      await apiClient.post<ApiResponse<void>>("/auth/resend-verify-otp", data)
    ).data;
  },

  async forgotPassword(data: ForgotPasswordRequest) {
    return (
      await apiClient.post<ApiResponse<void>>("/auth/forgot-password", data)
    ).data;
  },

  async resetPassword(data: ResetPasswordRequest) {
    return (await apiClient.post<ApiResponse<void>>("/auth/reset-password", data))
      .data;
  },

  async logout(): Promise<void> {
    try {
      await apiClient.post("/auth/logout");
    } finally {
      localStorage.removeItem("accessToken");
    }
  },

  async refreshToken(): Promise<TokenResponse> {
    const res = await apiClient.post<ApiResponse<TokenResponse>>(
      "/auth/refresh-token",
      null,
      { withCredentials: true }
    );
    return res.data.data;
  },

  async getCurrentUser(): Promise<User> {
    const res = await apiClient.get<ApiResponse<User>>("/auth/me");
    return res.data.data;
  },
};
